exa/core/data.py

```diff
@@ -4,11 +4,182 @@
 Data
 ########
 """
-from traitlets import Unicode, Integer, Float
+import importlib
+
+from traitlets import List, Unicode, Dict, Any
+from traitlets import validate, default, observe
+from traitlets import TraitError
 from traitlets.config import Configurable
+import pandas as pd
+import yaml
+
+import exa
+from exa.core.error import RequiredColumnError
+
+
+# TODO: reintroduce the automated getattr(attr, getter_attr())
+#       with _getter_prefix behavior on a
+#       base class somewhere that providers
+#       inherit from
+
+
+class Data(exa.Base, Configurable):
+    """An interface to separate data provider routing
+    logic and simplify managing multiple data concepts
+    in the container.
+    """
+    name = Unicode()
+    source = Any(allow_none=True)
+    source_args = List()
+    source_kws = Dict()
+    index = Unicode()
+    columns = List()
+    categories = Dict()
+
+    @validate('source')
+    def _validate_source(self, prop):
+        """source must implement __call__"""
+        if not callable(prop['value']):
+            raise TraitError("source must be callable")
+        return prop['value']
+
+    @observe('source')
+    def _observe_source(self, change):
+        """Update stored data if source changes. Nullify
+        related traits when source is nullified to reset
+        state of the Data object."""
+        if change.new is None:
+            self._data = None
+            self.source_args = []
+            self.sourcs_kws = {}
+
+    @validate('name')
+    def _validate_name(self, prop):
+        return prop['value'].lower()
+
+    @default('name')
+    def _default_name(self):
+        return self.__class__.__name__
+
+    @classmethod
+    def from_yml(cls, path):
+        with open(path, 'r') as f:
+            cfg = yaml.safe_load(f.read())
+        ver = cfg.pop('version', None)
+        if ver != 1:
+            raise Exception("version {} not supported".format(ver))
+        source = cfg.pop('source', None)
+        # Assume source is a path to a method_name or ClassName
+        # and is fully qualified by an importable module
+        if source is not None:
+            try:
+                *mod, obj = source.split('.')
+                mod = importlib.import_module('.'.join(mod))
+                # TODO : do we also support providing init
+                #        args and kwargs in case source is class?
```

Comment:
In the tests I see `source_args` and kwargs. Were those intended to be passed to the constructor of the class?

Reply:
Ahh yes this is something I mentioned in #167 perhaps we should move the design discussion to the Issue. I will post a response there.